bpo-29854: Fix segfault in call_readline()

nirs opened the pull request:
If history-length is set in .inputrc, and the history file is double the
history size (or more), history_get(N) returns NULL, and python
segfaults. Fix that by checking for NULL return value.

It seems that the root cause is incorrect handling of bigger history in
readline, but python should not segfault even if readline returns
unexpected value.

nirs committed:
bpo-29854: Fix segfault in call_readline()

If history-length is set in .inputrc, and the history file is double the
history size (or more), history_get(N) returns NULL, and python
segfaults. Fix that by checking for NULL return value.

It seems that the root cause is incorrect handling of bigger history in
readline, but python should not segfault even if readline returns
unexpected value.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

nirs commented:
I signed the CLA few years ago, but my github user name was missing in bpo.

vadmium commented on Modules/readline.c:
```diff
@@ -1347,15 +1347,17 @@ call_readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
-            line = (const char *)history_get(length)->line;
-        else
+                hist_ent = history_get(length);
+            line = (const char *)hist_ent ? hist_ent->line : "";
```
Why do you cast hist_ent? The original cast was added in Git revision 2525dc8, though I have doubts about the reason [avoiding a compiler warning when assigning (char *) to (const char *)]. It is better to avoid casts if possible; they can mask real errors and warnings.

serhiy-storchaka replied:
The cast is not needed. A compiler shouldn't emit a warning when assigning `char *` to `const char *`.

And more, I have doubts about the priority of the cast and the trinary operator. Does the above code is equivalent to `(const char *)(hist_ent ? hist_ent->line : "")` or to `((const char *)hist_ent) ? hist_ent->line : ""`?

nirs replied:
I think the correct cast would be:

    hist_ent ? (const char *)hist_ent->line : "";

But the cast is probably not needed, I kept it only to minimize changes which are not required to fix this issue.

serhiy-storchaka commented on Modules/readline.c:
```diff
@@ -1347,15 +1347,17 @@ call_readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
-        else
+                hist_ent = history_get(length);
+            line = (const char *)hist_ent ? hist_ent->line : "";
+        } else
```
`}` and `else` should be on separate lines.

nirs replied:
I tried to keep the current code style of this module, see for example write_history_file, read_history_file, read_init_file, setup_readline.

nirs committed:
Allow setting environment for readline tests

This enable testing custom readline configuration using the INPUTRC
environment variable, or passing arguments to the child process in a
clean way.

nirs committed:
bpo-29854: Add failing test that crash readline

readline segfaults on input() if the number of items in the history file
is equal or more to history size * 2.

This issue affects only GNU readline. When using libedit emulation
system history size option does not work.

nirs commented:
Changes in version 2:
- Add failing test (serhiy-storchaka)
- Remove unneeded cast in readline.c (vadmium)

nirs commented:
Version 3 adds the missing NEWS entry.

vadmium commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+import readline
+import sys
+
+history_file = "{}"
```
Might be safer to use _ascii_ or {!a} (_repr_ or {!r} on Python 2), in case the path has special characters; e.g. double quotes are allowed on Unix, Windows user profiles may be non-ASCII. Or you could change directory to the temp directory in the child process, or pass it as a CLI argument.

nirs replied:
The path is a temporary path, so it should not have quotes. I think a better way would be to use command line arguments or environment variables instead.

vadmium commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+
+            script = """
+import readline
+import sys
```
Is this needed?

nirs replied:
Leftover, will remove.

vadmium commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+            inputrc = os.path.join(temp_dir, "inputrc")
+            with io.open(inputrc, "wb") as f:
+                f.write(b"set history-size %d\n" % history_size)
+            env = os.environ.copy()
```
I think it would be safer and clearer to make the copy with _env = dict(os.environ)_. _Os.environ.copy_ is undocumented.

nirs replied:
I agree, will update in the next version.

berkerpeksag commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+            self.assertEqual(len(lines), history_size)
+            self.assertEqual(lines[-1].strip(), b"last input")
+        finally:
+            shutil.rmtree(temp_dir)
```
Can't we use `self.addCleanup(shutil.rmtree, temp_dir)` instead of this try..finally block?

serhiy-storchaka replied:
Or use `test.support.temp_dir()` or `test.support.temp_cwd()`.

nirs replied:
Good idea, will change.

berkerpeksag commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+        temp_dir = tempfile.mkdtemp()
+        try:
+            inputrc = os.path.join(temp_dir, "inputrc")
+            with io.open(inputrc, "wb") as f:
```
Did you use `io.open()` to make backporting to 2.7 easier?

serhiy-storchaka replied:
I don't think there is a significant difference between Python 2 builtin `open()` and `io.open()` here.

nirs replied:
Yes, code that works in both python 3 and 2 is my intent.

berkerpeksag commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+readline.read_history_file(history_file)
+input()
+readline.write_history_file(history_file)
+""".format(history_file)
```
Style nit: I'd rather use an f-string here.

nirs replied:
f-string does not work on python 2, creating extra work :-)

berkerpeksag commented on Misc/NEWS:
```diff
@@ -322,6 +322,9 @@ Core and Builtins
 -----------------
 
+- bpo-29854: Fix segfault in readline when using readline's history-size
+  option.
```
Please add "Patch by Nir Soffer." (and add two spaces after the full stop)

nirs replied:
Sure.

serhiy-storchaka commented on Lib/test/test_readline.py:
```diff
@@ -210,13 +212,49 @@ def display(substitution, matches, longest_match_length):
+        try:
+            inputrc = os.path.join(temp_dir, "inputrc")
+            with io.open(inputrc, "wb") as f:
+                f.write(b"set history-size %d\n" % history_size)
```
This is the first time I see using bytes formatting in tests! Since bytes formatting is supported in Python 3.5 this is okay.

nirs commented:
Changes in version 4:
- Avoid undocumented os.environ.copy (vadmium)
- Remove unneeded import (vadmium)
- Pass configuraion to child process using environment variables instead of formatting (vadmium, berkerpeksag)
- Use addCleanup instead of try-finally (berkerpeksag)
- Adding missing "Patch by" to news entry (berkerpeksag)

nirs commented:
Version 4 is ready for 25 days, it would be nice if someone can take a look :-)

berkerpeksag reviewed (commented):
The last version of the PR looks good to me, thank you!

I still prefer to use `open()` instead of `io.open()` (we are likely get a merge conflict in Misc/NEWS file when we backport it to 2.7 so we will need to do manual edit anyway and I don't think 2.7 is important to write slightly less idiomatic code in Python 3), but I'm going to left the ultimate decision to Martin and Serhiy.

serhiy-storchaka commented:
LGTM if use builtin `open()`.

nirs commented:
Changes in version 5:
- Use builtin open instead of io.open (io.open is required for backporting to 2.7).

nirs commented:
@serhiy-storchaka, I saw your comment about using test.support.temp_dir after I uploaded version 5.

See https://github.com/nirs/cpython/commit/45931f52dc3a9c15a207219d35ac5d5541a222b8 using it. I can squash it into the tests patch or send another pull request later if you like.

serhiy-storchaka commented:
@nirs, sorry, I added my comments few days ago, but forgot to publish them. When I approved the PR, they became visible.

I still think that it would be better to use `test.support.temp_dir()` (it is more robust for testing on Windows), but this is not critical, and I approved the PR in any case.

nirs commented:
@serhiy-storchaka are you waiting for another change?

berkerpeksag commented:
I think we need to move the NEWS entry into `Misc/NEWS.d/next/Library` now. You can use [blurb](https://github.com/python/core-workflow/tree/master/blurb) to create a NEWS file.

I liked https://github.com/nirs/cpython/commit/45931f52dc3a9c15a207219d35ac5d5541a222b8. Could you integrate it into this PR?

I can do the merging if Serhiy is busy with other PRs. Thanks! (and sorry for taking too long to merge this.)

serhiy-storchaka commented:
> @serhiy-storchaka are you waiting for another change?

No, I just waited for other reviewers in the case they have other comments. I left this on Berker.

nirs commented:
@berkerpeksag, ok, I will squash nirs/cpython@45931f5 into the test patch, and update the news item as you suggest.

nirs commented:
@berkerpeksag are you sure blurb is ready? see https://github.com/python/core-workflow/issues/153

nirs commented:
Changes in version 3:

- Integrate nirs/cpython@45931f5 into the tests patch (berkerpeksag)
- move the NEWS entry into Misc/NEWS.d/next/Library (berkerpeksag)

@berkerpeksag, @serhiy-storchaka please review.

berkerpeksag commented:
Thanks!

nirs commented:
@berkerpeksag, @serhiy-storchaka, @vadmium thanks for reviewing!

Do you want me to backport this to older versions?

berkerpeksag commented:
@nirs it would be better to fix the test failures reported at https://bugs.python.org/issue29854#msg297876 and https://bugs.python.org/issue29854#msg297877 before doing the backports. Do you have some time to take a look at those failures?

nirs commented:
@berkerpeksag sure, I'll take a look.
